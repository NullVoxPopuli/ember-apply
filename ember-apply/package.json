{
  "name": "ember-apply",
  "version": "2.2.1",
  "description": "automatic integration and configuration from the community",
  "repository": {
    "url": "https://github.com/NullVoxPopuli/ember-apply",
    "type": "https",
    "directory": "packages/ember-apply"
  },
  "main": "src/index.js",
  "exports": {
    ".": "./src/index.js",
    "./test-utils": "./src/test-utils.js"
  },
  "type": "module",
  "author": "NullVoxPopuli",
  "license": "MIT",
  "private": false,
  "bin": {
    "ember-apply": "src/cli/index.js"
  },
  "types": "./declarations/src/",
  "files": [
    "src/**/*",
    "declarations/**/*",
    "README.md"
  ],
  "scripts": {
    "prepack": "yarn build:types && cp ../README.md ./",
    "build:types": "tsc --build",
    "lint": "yarn lint:js",
    "lint:js": "eslint .",
    "lint:fix": "yarn lint:js --fix",
    "test": "vitest --coverage --no-watch --no-threads",
    "test:watch": "vitest --watch --no-threads"
  },
  "dependencies": {
    "@babel/core": "^7.16.7",
    "@babel/preset-env": "^7.16.8",
    "chalk": "^5.0.0",
    "ember-template-recast": "^6.1.2",
    "execa": "^6.0.0",
    "fs-extra": "^10.0.0",
    "jscodeshift": "^0.13.1",
    "latest-version": "^6.0.0",
    "ora": "^6.0.1",
    "pacote": "^13.0.0",
    "posthtml": "^0.16.5",
    "semver": "^7.3.5",
    "unified": "^10.1.1",
    "yargs": "^17.3.1",
    "yarn-workspaces-list": "^0.2.0"
  },
  "devDependencies": {
    "@nullvoxpopuli/eslint-configs": "^2.1.17",
    "@types/fs-extra": "^9.0.13",
    "@types/jscodeshift": "^0.11.3",
    "@types/node": "^17.0.8",
    "@types/pacote": "^11.1.3",
    "@types/semver": "^7.3.9",
    "@types/yargs": "^17.0.8",
    "@vitest/ui": "^0.6.0",
    "babel-eslint": "^10.1.0",
    "c8": "^7.11.0",
    "eslint": "^7.0.0",
    "eslint-plugin-decorator-position": "^4.0.1",
    "eslint-plugin-import": "^2.25.4",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^4.0.0",
    "prettier": "^2.5.1",
    "semantic-release": "^19.0.0",
    "semantic-release-monorepo": "^7.0.0",
    "typescript": "4.6.2",
    "vite": "^2.7.13",
    "vitest": "^0.5.0"
  },
  "engines": {
    "node": ">=16.0.0"
  },
  "volta": {
    "node": "16.14.0",
    "yarn": "1.22.17"
  }
}
