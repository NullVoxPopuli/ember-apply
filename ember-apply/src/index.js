// @ts-check
//
//
// Each of these re-exported modules points at an index.js file which
// defines the Public API for each namespace

export * as automation from './-private/automation/index.js';
export * as css from './-private/css/index.js';
export * as ember from './-private/ember/index.js';
export * as files from './-private/files/index.js';
export * as html from './-private/html/index.js';
export * as js from './-private/js/index.js';
export * as npm from './-private/npm/index.js';
export * as packageJson from './-private/package-json/index.js';
export * as project from './-private/project/index.js';
