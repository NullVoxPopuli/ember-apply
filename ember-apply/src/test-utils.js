import { execa } from 'execa';
import fs from 'fs/promises';
import os from 'os';
import path, { dirname } from 'path';
import { fileURLToPath } from 'url';

import { packageJson, project } from './index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

export async function newTmpDir() {
  const tmpDir = await fs.mkdtemp(
    path.join(os.tmpdir(), `ember-apply-${Date.now()}-`),
  );

  return tmpDir;
}

export async function newEmberApp() {
  let dir = await newTmpDir();

  await execa('ember', ['-v'], { cwd: dir });
  await execa('ember', ['new', 'test-app', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-app');
}

export async function newEmberAddon() {
  let dir = await newTmpDir();

  await execa('ember', ['-v'], { cwd: dir });
  await execa('ember', ['addon', 'test-addon', '--skip-npm'], {
    cwd: dir,
  });

  return path.join(dir, 'test-addon');
}

/**
 * @param {string} appPath the app to apply the applyable to
 * @param {string} applyablePath the path to the applyable
 */
export async function apply(appPath, applyablePath) {
  let cli = path.join(__dirname, 'cli/index.js');
  let target = applyablePath.startsWith('@')
    ? applyablePath
    : path.resolve(applyablePath);

  await execa('node', [cli, '--verbose', target], {
    cwd: appPath,
    stdio: process.env.DEBUG ? 'inherit' : undefined,
  });
}

/**
 * @param {string} appPath the path to git diff
 */
export async function diffSummary(appPath) {
  let { stdout } = await execa('git', ['diff', '--compact-summary'], {
    cwd: appPath,
  });

  return stdout;
}

/**
 * @typedef {object} DiffOptions
 * @property {boolean} [ignoreVersions]
 *
 * @param {string} appPath the path to git diff
 * @param {DiffOptions} [options]
 */
export async function diff(appPath, options = {}) {
  let { stdout } = await execa('git', ['diff', '--name-only'], {
    cwd: appPath,
  });

  let files = stdout.split('\n');

  let diff = '';

  for (let filePath of files) {
    if (!filePath) {
      continue;
    }

    if (filePath.endsWith('package.json') && options?.ignoreVersions) {
      let fileBuffer = await fs.readFile(path.join(appPath, filePath));
      let fileString = fileBuffer.toString();
      let json = JSON.parse(fileString);

      for (let section of [
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'engines',
        'volta',
      ]) {
        if (json[section]) {
          json[section] = Object.keys(json[section]);
        }
      }

      diff += `\n${filePath}\n${JSON.stringify(json, null, 2)}\n`;
      continue;
    }

    // Why is this not easier....
    // let normalCommand =
    //   `git diff ${filePath} ` +
    //   `| sed -n '/^---/!p' ` +
    //   `| sed -n '/^+++/!p' ` +
    //   `| sed -n '/^@@/!p' ` +
    //   `| sed -n '/^index /!p'`;

    let { stdout: fullDiff } = await execa(`git`, ['diff', filePath], {
      cwd: appPath,
    });

    let shortDiff = fullDiff
      .replace(/^---.+$/, '')
      .replace(/^\+\+\+.+$/, '')
      .replace(/index.+$/, '')
      .replace(/@@.+$/, '');

    diff += shortDiff + '\n';
  }

  return diff;
}

/**
 * Gets text content of the file all in one go
 * fs.readFile with the extra steps
 *
 * @param {string} filePath
 * @returns {Promise<string>} the file
 */
export async function readFile(filePath) {
  let file = await fs.readFile(filePath);

  return file.toString();
}

/**
 * Gets JSON content of the file all in one go
 * fs.readFile with the extra steps
 *
 * @param {string} filePath
 * @returns {Promise<object>} the file
 */
export async function readJson(filePath) {
  let text = await readFile(filePath);

  return JSON.parse(text);
}

/**
 * Creates a new monorepo with git in a tmp directory
 *
 * @param {string[]} workspaces list of workspace names
 */
export async function newMonorepo(workspaces) {
  let root = await newTmpDir();

  await execa('git', ['init'], { cwd: root });

  await fs.writeFile(
    path.join(root, 'package.json'),
    `{ "workspaces": [${workspaces.map((w) => `"${w}"`)}], "private": true }`,
  );
  await fs.writeFile(
    path.join(root, 'pnpm-workspace.yaml'),
    `packages:\n` + workspaces.map((w) => `- ${w}\n`).join(''),
  );

  for (let workspace of workspaces) {
    let safeName = workspace.replaceAll('/', '-');

    await fs.mkdir(path.join(root, `${workspace}`), { recursive: true });
    await fs.writeFile(
      path.join(root, `${workspace}/package.json`),
      `{ "name": "test-${safeName}" , "dependencies": {}, "version": "0.0.0" }`,
    );
  }

  return root;
}

/**
 * @param {string} root
 */
export async function readAllPackageJson(root) {
  let workspaces = await project.getWorkspaces(root);

  let manifests = [];

  // done serially, so order is consistent
  for (let workspace of workspaces) {
    let manifest = await packageJson.read(workspace);

    manifests.push(manifest);
  }

  return manifests;
}
