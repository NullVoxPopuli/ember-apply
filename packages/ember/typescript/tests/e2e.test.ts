import fs from 'node:fs/promises';
import path from 'node:path';

import { packageJson } from 'ember-apply';
import { apply, diff, diffSummary, newEmberApp } from 'ember-apply/test-utils';
import { execa } from 'execa';
import { describe, expect, it } from 'vitest';

import { default as typescript } from '../index.js';

describe('typescript', () => {
  it('default export exists', () => {
    expect(typeof typescript).toEqual('function');
  });

  describe('applying to an ember app', () => {
    it('works via CLI', async () => {
      let appLocation = await newEmberApp();

      let getManifest = () => packageJson.read(appLocation);

      // Don't currently support ember-data
      await packageJson.removeDevDependencies(['ember-data', 'ember-welcome-page'], appLocation);
      expect((await getManifest()).devDependencies).to.not.toHaveProperty('ember-data');
      expect((await getManifest()).devDependencies).to.not.toHaveProperty('ember-welcome-page');
      await execa('pnpm', ['install', '--no-frozen-lockfile'], { cwd: appLocation });

      // ember-page-title doesn't have types yet
      //   https://github.com/ember-cli/ember-page-title/pull/275
      await fs.writeFile(path.join(appLocation, 'app/templates/application.hbs'), `hello`);

      await apply(appLocation, typescript.path);
      expect((await getManifest()).scripts).to.toHaveProperty('lint:types');

      expect(await diffSummary(appLocation)).toMatchSnapshot();
      expect(
        await diff(appLocation, { ignoreVersions: true })
      ).toMatchSnapshot();

      let install = await execa('pnpm', ['install', '--no-frozen-lockfile'], { cwd: appLocation });

      expect(install.exitCode, 'pnpm install').toBe(0);

      let lintTypes = await execa('pnpm', ['lint:types'], { cwd: appLocation });

      expect(lintTypes.exitCode, 'lint:types').toBe(0);
    });
  });
});
