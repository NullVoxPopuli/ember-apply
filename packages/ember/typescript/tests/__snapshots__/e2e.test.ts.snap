// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typescript > applying to an ember app > works via CLI 1`] = `
" app/templates/application.hbs |  8 +-------
 ember-cli-build.js            |  4 +++-
 package.json                  | 21 ++++++++++++++++++---
 3 files changed, 22 insertions(+), 11 deletions(-)"
`;

exports[`typescript > applying to an ember app > works via CLI 2`] = `
"diff --git a/app/templates/application.hbs b/app/templates/application.hbs
index 0eacef7..b6fc4c6 100644
--- a/app/templates/application.hbs
+++ b/app/templates/application.hbs
@@ -1,7 +1 @@
-{{page-title \\"TestApp\\"}}
-
-{{! The following component displays Ember's default welcome message. }}
-<WelcomePage />
-{{! Feel free to remove this! }}
-
-{{outlet}}
\\\\ No newline at end of file
+hello
\\\\ No newline at end of file
diff --git a/ember-cli-build.js b/ember-cli-build.js
index ed991bd..8557327 100644
--- a/ember-cli-build.js
+++ b/ember-cli-build.js
@@ -4,7 +4,9 @@ const EmberApp = require('ember-cli/lib/broccoli/ember-app');
 
 module.exports = function (defaults) {
   const app = new EmberApp(defaults, {
-    // Add options here
+    \\"ember-cli-babel\\": {
+      enableTypeScriptTransform: true
+    }
   });
 
   return app.toTree();

package.json
{
  \\"name\\": \\"test-app\\",
  \\"version\\": \\"0.0.0\\",
  \\"private\\": true,
  \\"description\\": \\"Small description for test-app goes here\\",
  \\"repository\\": \\"\\",
  \\"license\\": \\"MIT\\",
  \\"author\\": \\"\\",
  \\"directories\\": {
    \\"doc\\": \\"doc\\",
    \\"test\\": \\"tests\\"
  },
  \\"scripts\\": {
    \\"build\\": \\"ember build --environment=production\\",
    \\"lint\\": \\"concurrently \\\\\\"npm:lint:*(!fix)\\\\\\" --names \\\\\\"lint:\\\\\\"\\",
    \\"lint:css\\": \\"stylelint \\\\\\"**/*.css\\\\\\"\\",
    \\"lint:css:fix\\": \\"concurrently \\\\\\"npm:lint:css -- --fix\\\\\\"\\",
    \\"lint:fix\\": \\"concurrently \\\\\\"npm:lint:*:fix\\\\\\" --names \\\\\\"fix:\\\\\\"\\",
    \\"lint:hbs\\": \\"ember-template-lint .\\",
    \\"lint:hbs:fix\\": \\"ember-template-lint . --fix\\",
    \\"lint:js\\": \\"eslint . --cache\\",
    \\"lint:js:fix\\": \\"eslint . --fix\\",
    \\"start\\": \\"ember serve\\",
    \\"test\\": \\"concurrently \\\\\\"npm:lint\\\\\\" \\\\\\"npm:test:*\\\\\\" --names \\\\\\"lint,test:\\\\\\"\\",
    \\"test:ember\\": \\"ember test\\",
    \\"lint:types\\": \\"glint\\"
  },
  \\"devDependencies\\": [
    \\"@babel/core\\",
    \\"@babel/eslint-parser\\",
    \\"@babel/plugin-proposal-decorators\\",
    \\"@ember/optional-features\\",
    \\"@ember/string\\",
    \\"@ember/test-helpers\\",
    \\"@glimmer/component\\",
    \\"@glimmer/tracking\\",
    \\"@glint/core\\",
    \\"@glint/environment-ember-loose\\",
    \\"@glint/environment-ember-template-imports\\",
    \\"@glint/template\\",
    \\"@tsconfig/ember\\",
    \\"@types/qunit\\",
    \\"@types/rsvp\\",
    \\"@typescript-eslint/eslint-plugin\\",
    \\"@typescript-eslint/parser\\",
    \\"broccoli-asset-rev\\",
    \\"concurrently\\",
    \\"ember-auto-import\\",
    \\"ember-cli\\",
    \\"ember-cli-app-version\\",
    \\"ember-cli-babel\\",
    \\"ember-cli-clean-css\\",
    \\"ember-cli-dependency-checker\\",
    \\"ember-cli-htmlbars\\",
    \\"ember-cli-inject-live-reload\\",
    \\"ember-cli-sri\\",
    \\"ember-cli-terser\\",
    \\"ember-fetch\\",
    \\"ember-load-initializers\\",
    \\"ember-modifier\\",
    \\"ember-page-title\\",
    \\"ember-qunit\\",
    \\"ember-resolver\\",
    \\"ember-source\\",
    \\"ember-template-lint\\",
    \\"eslint\\",
    \\"eslint-config-prettier\\",
    \\"eslint-plugin-ember\\",
    \\"eslint-plugin-n\\",
    \\"eslint-plugin-prettier\\",
    \\"eslint-plugin-qunit\\",
    \\"loader.js\\",
    \\"prettier\\",
    \\"qunit\\",
    \\"qunit-dom\\",
    \\"stylelint\\",
    \\"stylelint-config-standard\\",
    \\"stylelint-prettier\\",
    \\"tracked-built-ins\\",
    \\"typescript\\",
    \\"webpack\\"
  ],
  \\"engines\\": [
    \\"node\\"
  ],
  \\"ember\\": {
    \\"edition\\": \\"octane\\"
  },
  \\"pnpm\\": {
    \\"overrides\\": {
      \\"@glimmer/manager\\": \\">= 0.84.3\\",
      \\"@glimmer/validator\\": \\">= 0.84.3\\"
    }
  }
}
"
`;
